### DISCUSSION TITLE
Python Data Fetcher

### DISCUSSION THREAD

**Discussion by CaffeineLab** [ercot/api-specs#49] [Show and tell]
Just a heads up that I've added an ERCOT interface to the python-isodata project at [https://github.com/CaffeineLab/isodata](https://github.com/CaffeineLab/isodata)

Currently looks like this, but check out the repo as changes are still being made.
```
from pathlib import Path
import pandas as pd
from isodata.sessions import Session

ercot = Session('ercot_public')
ercot.authorize(username=creds['credentials']['ercot_public_api']['username'],
                password=creds['credentials']['ercot_public_api']['password'],
                primary_key=creds['credentials']['ercot_public_api']['primary_key'],
                auth_url=creds['credentials']['ercot_public_api']['auth_url'])

# Fetch 'SCED Shadow Prices and Binding Transmission Constraints'
emil = 'NP6-86-CD'

# Fetch the list of documents on the first page
report_list, count = ercot.fetch_listing(emil_id=emil, page=1)

# Fetch the first file and save it locally, the 2nd element
# in the tuple is the link to the report.
data_file = ercot.fetch_url(report_list[0][2], Path('path/to/save/report'))

# Do what you want with the data in the file from here.
# Beware of zips with multiple files, or data_files that 
# are just plain .csv files.  An example for NP6-86-CD data is:
df = pd.read_csv(data_file, compression='zip')
```

**Comment by mackermann-ercot** [comment]
Nice!

Regards,
mackermann-ercot

**Reply by CaffeineLab** [reply]
Thanks... and ... .THANKS!  The new API is quite nice and much easier to work with then the xml packages of the past.

**Comment by nealbgit** [comment]
Hey CaffeineLab - Nice project!

Question, in your code above, I'm using the standard Auth_url, but that doesn't seem to be the correct one.

I'm getting errors:
2024-10-19 14:20:16.853 | ERROR    | isodata.ercot.connector:get_token:126 - 400 Client Error: Bad Request for url: https://ercotb2c.b2clogin.com/ercotb2c.onmicrosoft.com/B2C_1_PUBAPI-ROPC-FLOW/oauth2/v2.0/token
2024-10-19 14:20:16.855 | ERROR    | isodata.connector:authorize:20 - No token retrieved

Can you point me in the right direction?

**Reply by CaffeineLab** [reply]
Yeah - it's a bit nasty, but the the url needs to be the WHOLE string as below:

`https://ercotb2c.b2clogin.com/ercotb2c.onmicrosoft.com/B2C_1_PUBAPI-ROPC-FLOW/oauth2/v2.0/token?username={username}&password={password}&grant_type=password&scope=openid+fec253ea-0d06-4272-a5e6-b478baeecd70+offline_access&client_id=fec253ea-0d06-4272-a5e6-b478baeecd70&response_type=id_token
`

In the isodata library it will use the usname and password provided to convert the tokens.  I'll put some more documentation on it up there later.  Also in my example above...  

`data_file = ercot.fetch_url(report_list[0][2], 'path/to/save/report')`

should pass a pathlib.Path, not a string.  

`data_file = ercot.fetch_url(report_list[0][2], pathlib.Path('path/to/save/report'))`

Let me know if that helps - if not, I'll help sort it out.

**Reply by nealbgit** [reply]
That worked - thanks.  

To make sure you have a full working example, your imports should be:

from isodata.sessions import Session
import pathlib
import pandas as pd

And for Windows folks out there, the pathlib line should look something like this:
data_file = ercot.fetch_url(report_list[0][2], pathlib.Path('C:/Temp/Reports'))

**Comment by nealbgit** [comment]
@CaffeineLab -

One other question for you - does your Python package handle secure downloads requiring client certificates?
For example, can you pass .PFX file and password params for a location such as:

QSE Load Ratio Share
https://mis.ercot.com/secure/data-products/markets/day-ahead-market?id=NP4-34-CD

Thanks.

**Reply by CaffeineLab** [reply]
Not yet.  But I can probably get that working later today...

**Reply by nealbgit** [reply]
That'd be great.  Glad to test for you!

**Reply by CaffeineLab** [reply]
I pushed up a Q&D just now that's pulling the list and files.  Also - moving the convseration to [https://github.com/CaffeineLab/isodata/discussions](https://github.com/CaffeineLab/isodata/discussions)
